fix(react-query/HydrationBoundary): prevent unnecessary refetch during hydration

sukvvon opened the pull request:
## 🎯 Changes

Fixes #9610

When using `HydrationBoundary` with existing queries in the cache, an unnecessary refetch was triggered during hydration even though fresh data was about to be hydrated from the server.

### Root Cause
- `useQuery`'s `onSubscribe` runs before `HydrationBoundary`'s `useEffect` completes hydration
- The query appears stale (before hydration), triggering an unnecessary refetch

### Solution
- Added `IsHydratingContext` (React Context) to track queries pending hydration
- Mark existing queries as pending in `useMemo` (before children render)
- Skip refetch in `queryObserver.#executeFetch` if query is pending hydration and data is fresh
- Exception: `refetchOnMount: 'always'` still triggers refetch (user explicitly wants fresh data)
- Exception: Stale data (based on `staleTime`) triggers refetch (cached markup scenario)
- Clear pending flag in `useEffect` after hydration completes

### Why React Context instead of global state?
- Supports multiple `QueryClient` instances (micro-frontends, testing)
- Proper isolation between different `HydrationBoundary` components
- Follows React patterns for state management

### Files Changed
- `packages/query-core/src/queryObserver.ts` - Check pending hydration and stale state in `#executeFetch`
- `packages/query-core/src/types.ts` - Add `_isHydrating` option type
- `packages/react-query/src/HydrationBoundary.tsx` - Mark/clear pending hydration queries with Context
- `packages/react-query/src/IsHydratingProvider.ts` - New Context provider for hydration state
- `packages/react-query/src/useBaseQuery.ts` - Set `_isHydrating` flag from Context
- `packages/react-query/src/useQueries.ts` - Set `_isHydrating` flag from Context
- `packages/react-query/src/index.ts` - Export `useIsHydrating` and `IsHydratingProvider`
- `packages/react-query/src/__tests__/HydrationBoundary.test.tsx` - Add 26 test cases
- `docs/framework/react/guides/ssr.md` - Document new behavior
- `docs/framework/react/reference/hydration.md` - Document new behavior

## ✅ Checklist

- [x] I have followed the steps in the [Contributing guide](https://github.com/TanStack/query/blob/main/CONTRIBUTING.md).
- [x] I have tested this code locally with `pnpm run test:pr`.

## 🚀 Release Impact

- [x] This change affects published code, and I have generated a [changeset](https://github.com/changesets/changesets/blob/main/docs/adding-a-changeset.md).
- [ ] This change is docs/CI/dev-only (no release).

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added a hydration tracking hook and provider to expose pending hydration state.

* **Bug Fixes**
  * Prevents unnecessary refetches during hydration when hydrated data is fresh.

* **Documentation**
  * Expanded hydration docs with server memory guidance and Next.js rewrite caveats; guidance on refetch behavior and staleTime.

* **Tests**
  * Extended hydration test suite with many scenarios to validate correct hydration/refetch behavior.

* **Chores**
  * Patch version bumps recorded.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

sukvvon committed:
fix(react-query/HydrationBoundary): prevent unnecessary refetch during hydration

sukvvon committed:
Merge branch 'main' into fix/hydration-boundary-double-fetch

sukvvon committed:
chore(.changeset): change version bump from 'patch' to 'minor'

sukvvon committed:
refactor(react-query/HydrationBoundary): add useEffect cleanup and extract 'clearPendingQueries' function

sukvvon committed:
Merge branch 'main' into fix/hydration-boundary-double-fetch

sukvvon committed:
Merge branch 'main' into fix/hydration-boundary-double-fetch

sukvvon committed:
Merge branch 'main' into fix/hydration-boundary-double-fetch

sukvvon committed:
Merge branch 'main' into fix/hydration-boundary-double-fetch

sukvvon committed:
chore(.changeset): change version bump from 'minor' to 'patch'

sukvvon committed:
Merge branch 'main' into fix/hydration-boundary-double-fetch

sukvvon committed:
Merge branch 'main' into fix/hydration-boundary-double-fetch

sukvvon committed:
Merge branch 'main' into fix/hydration-boundary-double-fetch

sukvvon committed:
Merge branch 'main' into fix/hydration-boundary-double-fetch

sukvvon committed:
Merge branch 'main' into fix/hydration-boundary-double-fetch

tarasvarshava commented:
Thanks a lot for this. I hope it gets merged soon 🤞

sukvvon committed:
Merge branch 'main' into fix/hydration-boundary-double-fetch

sukvvon committed:
Merge branch 'main' into fix/hydration-boundary-double-fetch

Ephem commented on docs/framework/react/reference/hydration.md:
```diff
@@ -114,6 +114,8 @@ function App() {
 
 > Note: Only `queries` can be dehydrated with an `HydrationBoundary`.
 
+> Note: `HydrationBoundary` automatically prevents unnecessary refetching during hydration. Queries being hydrated will not trigger a refetch on mount, unless `refetchOnMount` is explicitly set to `'always'`.
```
Thank you for this PR! I've only had time to skim it so far and it will take some time before I can dig in properly, but I wanted to provide some quick feedback here.

We also need to account for the case where there is some time in between fetching on the server, and hydrating on the client. This commonly happens when the markup is cached, in which case the hydration can happen _days_ after the fetch on the server. In this situation, we do want to render with the stale data initially (this is what the SSR pass did and we need to match it), but we also want to immediately refetch the data because it's stale.

Ephem commented on packages/query-core/src/hydration.ts:
```diff
@@ -14,6 +14,11 @@ import type { QueryClient } from './queryClient'
 
+// WeakSet to track queries that are pending hydration
+// Used to prevent double-fetching when HydrationBoundary defers hydration to useEffect
+export const pendingHydrationQueries: WeakSet<Query<any, any, any, any>> =
```
I don't think we can use a global singleton to keep track of this, what if there are several `queryClient`s? In that case a query might just be about to hydrate into _one_ of these, not all of them.

Annoying as that is, I think we need to use a context for this. Note that there is some prior art with `isRestoring`.

sukvvon committed:
refactor(react-query): replace global 'pendingHydrationQueries' with React Context

autofix-ci[bot] committed:
ci: apply automated fixes

sukvvon committed:
fix(query-core): refetch stale data during hydration for cached markup scenario

sukvvon committed:
docs(react): update hydration docs to reflect stale data refetch behavior

sukvvon committed:
refactor(react-query/IsHydratingProvider): simplify context definition

sukvvon committed:
refactor(react-query): simplify hydration context from '{ current: Set }' to 'Set'

sukvvon committed:
Merge branch 'main' into fix/hydration-boundary-double-fetch

sukvvon committed:
Merge branch 'main' into fix/hydration-boundary-double-fetch

sukvvon committed:
ci: trigger CI

sukvvon committed:
Merge branch 'main' into fix/hydration-boundary-double-fetch

sukvvon committed:
Merge branch 'main' into fix/hydration-boundary-double-fetch

sukvvon commented:
@Ephem Could you review this PR?

sukvvon committed:
Merge branch 'main' into fix/hydration-boundary-double-fetch

sukvvon committed:
Merge branch 'main' into fix/hydration-boundary-double-fetch

sukvvon committed:
Merge branch 'main' into fix/hydration-boundary-double-fetch

sukvvon committed:
Merge branch 'main' into fix/hydration-boundary-double-fetch

sukvvon committed:
Merge branch 'main' into fix/hydration-boundary-double-fetch

sukvvon committed:
Merge branch 'main' into fix/hydration-boundary-double-fetch

Ephem commented on packages/react-query/src/HydrationBoundary.tsx:
```diff
@@ -50,62 +51,92 @@ export const HydrationBoundary = ({
-        }
+  // Create a mutable ref object to hold pending query hashes
+  // This allows us to add/remove hashes without triggering re-renders
+  const hydratingQueriesRef = React.useRef<Set<string>>(new Set())
```
I don't think this can be a ref unfortunately, it needs to be a state. It's being read in renders further down in context consumers, and it's not safe to read refs in render.

The reason is that with concurrent features, we want this to reflect the correct queries being hydrated _in this fork of the world_, we don't want to have the same state for the different concurrent realities so to speak.

Ephem commented on packages/query-core/src/queryObserver.ts:
```diff
@@ -96,7 +96,25 @@ export class QueryObserver<
+        resolvedRefetchOnMount !== 'always' &&
+        !isStale(this.#currentQuery, this.options)
+
+      if (
+        shouldFetchOnMount(this.#currentQuery, this.options) &&
+        !shouldSkipFetchForHydration
+      ) {
```
There's also logic in `createResult` that needs to account for this. That's also used to determine what state the query _will be in_, the optimistic result that the query hooks see on first render, before this is called and the actual fetch (possibly) happens.

This gets into some other tricky territory (where we might already have other bugs tbh), which is that the first render needs to match the render on the server. If the data was not stale when server rendering, but is now stale, we need to:

* Do the first render as if it's not stale to avoid hydration mismatches
* Still trigger a refetch and update the query _after_ the first render

Ephem reviewed (commented):
I'm so sorry for the radio silence here, I've been in the middle of a long distance move and had to step away from OSS for a few months, I'm still pretty busy IRL but I'm now easing back into it.

I appreciate this contribution, it's trying to tackle something pretty gnarly. I'm not at all happy about the current state of RQ when it comes to these things.

The long term fix here is even bigger, which is to put the actual queries that are being hydrated on a context. The hooks would read from that context and kind of "patch" the current store state with the query state from the context in case it's newer. This would mean we better support concurrent features, as even if the query already exists in the store, the concurrent render would still immediately see the new data, which is not the case today. This approach would also fix a few other bugs we have.

I need to get back into the code a bit more before I can fully evaluate if this PR might be a pragmatic middle step that builds towards that "full" fix, but I think so.

sukvvon committed:
Merge branch 'main' into fix/hydration-boundary-double-fetch

sukvvon committed:
fix(react-query/HydrationBoundary): derive pending hydration set in render instead of mutating a ref

sukvvon committed:
fix(query-core/queryObserver): match the server on the first hydrating render by not showing fetching state

sukvvon commented:
@Ephem First of all, my apologies for the long delay in following up here — thank you for the detailed review, it was really helpful in getting this right.

I've addressed both points:

**1. Pending hydration state via context, not a mutated ref** (your comment on `HydrationBoundary.tsx`)

You were right that a ref read in render isn't safe under concurrent features. Instead of mutating a shared `Set`, the pending set is now derived purely in render from `(client, state)`, so each render/fork carries its own immutable snapshot — no shared mutable object, nothing read from a ref during render. Consumers still read `.has(hash)`, but the value they read is a fresh per-render `Set`.

A nice consequence is that clearing is now automatic: once the deferred hydration lands and the cache's `dataUpdatedAt` matches the dehydrated data, the query is no longer classified as "newer", so it drops out of the pending set on the next render — no manual cleanup, and it stays correct for a late-mounting observer of the same key.

**2. `createResult` first render matching the server** (your comment on `queryObserver.ts`)

I reproduced the mismatch you described: when the server rendered non-stale but the client is now stale, the optimistic first render flipped to `isFetching: true`, which the server never rendered (dehydrated queries are always idle). The fix suppresses the optimistic fetch state while `_isHydrating`, so the first render matches the server regardless of staleness, and the actual refetch still fires afterwards in `onSubscribe` (unchanged) when the data is stale.

I've added tests covering both the pending-flag clearing and the first-render-not-fetching behavior. Would appreciate another look when you have time 🙏

sukvvon committed:
Merge branch 'main' into fix/hydration-boundary-double-fetch

sukvvon committed:
Merge branch 'main' into fix/hydration-boundary-double-fetch
